### PR TITLE
MLDBFB-474 add an option to the transform procedure to specify determi…

### PR DESCRIPTION
**Pull request by steadtler**
…nistic ordering or not. Deterministic is slower.

### CODE REVIEW

**Comment by FinchPowers** [comment]
+1, mais il faudrait que @nicolaskruchten donne son accord à ce paramètre.

**Comment by nicolaskruchten** [comment]
Hmmm I'm not sure I understand the need for such a parameter vs just using `ORDER BY` in the query... what's the downside of just hardcoding this to false all the time? and letting the user use order-by if they care?

File: plugins/sql_functions.cc [hunk 1]
```diff
@@ -611,7 +614,7 @@ run(const ProcedureRunConfig & run,
                          *runProcConf.inputData.stm->where,
                          runProcConf.inputData.stm->orderBy,
                          { runProcConf.inputData.stm->rowName })
-            .execute({recordRowInOutputDataset, true /*processInParallel*/},
+            .execute({recordRowInOutputDataset, !runProcConf.deterministicOrdering /*processInParallel*/},
```
Comment (guyd) [inline plugins/sql_functions.cc]: It was already executed in parallel.  The flag will only allow it to go slower.  That does not seem to fit the JIRA description!

**Comment by steadtler** [comment]
Ok! Lets explain this:

-Allowing output in parallel is the fastest, but the rows in the output dataset will be in a *different* order everytime you run it, since it depends on threading.
-Disallowing output in parallel will give the output in an *arbitrary* order, but it will always be the same order if you run it several times -> deterministic order.
-Having the user put an order by will give the result in a *specific* order, and it is the slowest.

1st option takes 20s, second option 1m15s (because some cores are starved), last option over an hour (according to Mich, I didnt test it) on Mich's dataset.

**Comment by nicolaskruchten** [comment]
Given that the target here is a dataset and not a query result, I'm ok with
the output of a transform to be nondeterministic always if it's fastest,
instead of making the user choose.

On Wednesday, 27 April 2016, Mathieu Marquis Bolduc <
notifications@github.com> wrote:

> Ok! Lets explain this:
>
> -Allowing output in parallel is the fastest, but the rows in the output
> dataset will be in a *different* order everytime you run it, since it
> depends on threading.
> -Disallowing output in parallel will give the output in an *arbitrary*
> order, but it will always be the same order if you run it several times ->
> deterministic order.
> -Having the user put an order by will give the result in a *specific*
> order, and it is the slowest.
>
> 1st option takes 20s, second option 1m15s (because some cores are
> starved), last option over an hour (according to Mich, I didnt test it) on
> Mich's dataset.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mldbai/mldb/pull/328#issuecomment-215066811>
>


-- 
Nicolas Kruchten, Eng  |  Staff Software Engineer at Datacratic  |
http://datacratic.com/

**Pull request closed without merging**